File: cpp/zeromq-client.cpp
#include <zmq.hpp>
#include <string>
#include <iostream>

#ifndef _WIN32
#include <unistd.h>
#else
#include <windows.h>
#define sleep(n)    Sleep(n)
#endif

void process(zmq::socket_t* socket, std::string implementation, int m, int n) {
  std::cout << "Computing ack (" << m << ")(" << n << ")" << std::endl;

  std::ostringstream payload;
  payload << "{ \"fn\": \"ack\", \"m\": " << m << ",\"n\": " << n << " }";
  std::string json(payload.str());

  zmq::message_t request (json.length());
  void *data = request.data();

  memcpy (data, json.c_str(), json.length());

  std::cout << "(" << implementation << ") Computing ack (" << m << ")(" << n << ")" << std::endl;
  socket->send(request, zmq::send_flags::none);

  zmq::message_t raw_reply;
  socket->recv (raw_reply, zmq::recv_flags::none);

  char *my_reply = (char*) strndup((char*)raw_reply.data(), raw_reply.size());

  std::cout << "(" << implementation << ") Received result: " << my_reply << std::endl;
}

int main () {
  std::cout << "Connecting to the AckFibFac Crystal Server...";
  zmq::context_t context_crystal (1);
  zmq::socket_t socket_crystal (context_crystal, ZMQ_REQ);
  socket_crystal.connect("tcp://localhost:6666");
  std::cout << "connected!" << std::endl;

  std::cout << "Connecting to the AckFibFac Ruby Server...";
  zmq::context_t context_ruby (1);
  zmq::socket_t socket_ruby (context_ruby, ZMQ_REQ);
  socket_ruby.connect ("tcp://localhost:5555");
  std::cout << "connected!" << std::endl;


  while(1) {
    int n = rand() % 3 + 1;
    int m = rand() % 3 + 1;

    process(&socket_ruby, "Ruby", m, n);
    process(&socket_crystal, "Crystal", m, n);
  }
}

File: ruby/zeromq-client.rb
# require 'bundler/inline'
require 'bundler/setup'

# gemfile do
#   source 'https://rubygems.org'

#   gem 'ffi-rzmq'
# end

require 'ffi-rzmq'
require 'json'
require 'benchmark'

context_crystal = ZMQ::Context.new

puts "Connecting to the AckFibFac Crystal Server..."
requester_crystal = context_crystal.socket(ZMQ::REQ)
requester_crystal.connect("tcp://localhost:6666")

context_ruby = ZMQ::Context.new

puts "Connecting to the AckFibFac Ruby Server..."
requester_ruby = context_ruby.socket(ZMQ::REQ)
requester_ruby.connect("tcp://localhost:5555")

loop do
  n = rand(4) + 1
  m = rand(4) + 1
  request = {fn: "ack", m: m, n: n}

  reply = ''
  elapsed_crystal = Benchmark.realtime do
    puts "Crystal: Computing #{request[:fn]}(#{m})(#{n})"
    requester_crystal.send_string request.to_json

    requester_crystal.recv_string(reply)
  end

  puts "Crystal (#{elapsed_crystal}): #{request[:fn]}(#{m})(#{n}): #{reply}"

  reply = ''
  elapsed_ruby = Benchmark.realtime do
    puts "Ruby: Computing #{request[:fn]}(#{m})(#{n})"
    requester_ruby.send_string request.to_json

    requester_ruby.recv_string(reply)
  end
  puts "Ruby (#{elapsed_ruby}): #{request[:fn]}(#{m})(#{n}): #{reply}"

  sleep 1
end
